### PR TITLE
drop runc-dmz solution according to overlay solution

### PR DESCRIPTION
**Pull request by lifubang**
Because we have the overlay solution, we can drop runc-dmz binary solution since it has too many limitations.

The original post is here: https://github.com/opencontainers/runc/issues/4450#issuecomment-2437088079

Because of there are many commits about runc-dmz binary solution, so doing git revert action is very hard, let's drop these code line by line.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -106,15 +106,13 @@ make BUILDTAGS=""
 | Build Tag     | Feature                               | Enabled by Default | Dependencies        |
 |---------------|---------------------------------------|--------------------|---------------------|
 | `seccomp`     | Syscall filtering using `libseccomp`. | yes                | `libseccomp`        |
-| `!runc_nodmz` | Reduce memory usage for CVE-2019-5736 protection by using a small C binary, [see `memfd-bind` for more details][contrib-memfd-bind]. `runc_nodmz` disables this **experimental feature** and causes runc to use a different protection mechanism which will further increases memory usage temporarily during container startup. To enable this feature you also need to set the `RUNC_DMZ=true` environment variable. | yes ||
 
 The following build tags were used earlier, but are now obsoleted:
  - **nokmem** (since runc v1.0.0-rc94 kernel memory settings are ignored)
  - **apparmor** (since runc v1.0.0-rc93 the feature is always enabled)
  - **selinux**  (since runc v1.0.0-rc93 the feature is always enabled)
```
Comment (AkihiroSuda) [inline README.md:113]: This historical list has to be updated for DMZ

File: contrib/cmd/memfd-bind/README.md [hunk 2]
```diff
@@ -1,8 +1,8 @@
 ## memfd-bind ##
 
-`runc` normally has to make a binary copy of itself (or of a smaller helper
-binary called `runc-dmz`) when constructing a container process in order to
-defend against certain container runtime attacks such as CVE-2019-5736.
+`runc` normally has to make a binary copy of itself when constructing a
+container process in order to defend against certain container runtime attacks
+such as CVE-2019-5736.
```
Comment (cyphar) [inline contrib/cmd/memfd-bind/README.md:5]: Ah, I forgot to document how the new overlay mode works in #4448. I'll open a separate PR for that.

File: libcontainer/container_linux.go [hunk 3]
```diff
@@ -474,54 +472,21 @@ func (c *Container) includeExecFifo(cmd *exec.Cmd) error {
...
 	if err != nil {
 		return nil, err
 	}
 
-	// Make sure we use a new safe copy of /proc/self/exe or the runc-dmz
+	// Make sure we use a new safe copy of /proc/self/exe
 	// binary each time this is called, to make sure that if a container
 	// manages to overwrite the file it cannot affect other containers on the
 	// system. For runc, this code will only ever be called once, but
 	// libcontainer users might call this more than once.
```
Comment (cyphar) [inline libcontainer/container_linux.go:485]: Can you re-wrap this comment? `gw` in vim (with `:set tw=80`).

**Review by cyphar — COMMENTED** [review]
One minor nit.

File: README.md (hunk 1 above)
Comment (lifubang) [reply README.md:114]: Has added!

File: libcontainer/container_linux.go (hunk 3 above)
Comment (lifubang) [reply libcontainer/container_linux.go]: Has changed.

File: README.md [hunk 4]
```diff
@@ -106,15 +106,14 @@ make BUILDTAGS=""
 | Build Tag     | Feature                               | Enabled by Default | Dependencies        |
 |---------------|---------------------------------------|--------------------|---------------------|
 | `seccomp`     | Syscall filtering using `libseccomp`. | yes                | `libseccomp`        |
-| `!runc_nodmz` | Reduce memory usage for CVE-2019-5736 protection by using a small C binary, [see `memfd-bind` for more details][contrib-memfd-bind]. `runc_nodmz` disables this **experimental feature** and causes runc to use a different protection mechanism which will further increases memory usage temporarily during container startup. To enable this feature you also need to set the `RUNC_DMZ=true` environment variable. | yes ||
 
 The following build tags were used earlier, but are now obsoleted:
+ - **runc_nodmz** (since runc v1.2.1 runc dmz binary is dropped)
```
Comment (AkihiroSuda) [inline README.md:111]: (Considering the amount of the changes I was wondering if this is going to be v1.3.0, but probably safe to cherrypick to v1.2.1, as dmz was experimental and opt-in)

Comment (AkihiroSuda) [reply README.md:111]: Probably https://github.com/opencontainers/runc/blob/main/docs/experimental.md should be updated to reflect the history

**Review by AkihiroSuda — APPROVED** [review]

**Review by cyphar — APPROVED** [review]

File: contrib/cmd/memfd-bind/README.md (hunk 2 above)
Comment (frezbo) [reply contrib/cmd/memfd-bind/README.md:5]: this is more of a discussion, so with the overlay change i guess memfd-bind is no longer needed? :thinking:

Comment (cyphar) [reply contrib/cmd/memfd-bind/README.md:5]: This is being discussed in https://github.com/opencontainers/runc/issues/4450#issuecomment-2448933143. Short answer: it has fewer upsides now and there is a fairly strong case for removing it.

**Pull request merged**